README.md
# Nioca

**CAUTION:**  
There has not been any third party security audit for this project.  
Use this software at your own risk!

**INFO:**  
This project is currently pre v1.0 and not intended for real production use yet.  
Things might change between versions until the API has been stabilized.  
However, it is working for testing and smaller setups so far.

## Project State

Even though I am using it already for my own DEV environments, this project is still in an early phase.  
The UI is not yet "prettified" and some parts and functionalities are completely missing. The basics are there and
are working fine so far. Almost everything has been prepared for the future features and options, even when they might
not be implemented yet.  
v0.4.0 is the first public release in this open source repository, but it has been used for testing and development
beforehand behind closed doors already. There is an independent `nioca-client` which can be used to easily fetch
certificates on external clients, which is especially useful for SSH. This client can be used as a standalone CLI tool
and therefore with simple system tasks on any machine, and it has a library implementation to be used from Rust apps
easily. This client is not yet public, but will get its own repo in the near future. Some work must be done before it
can be made public as well.

## What it is

Nioca is a Certificate Authority for TLS / X509 and SSH certificates.  
Its goal is to have a minimal footprint while being fast and efficient at the same time.

There has not been a third party security audit yet, but a lot of design has gone into the base setup and foundation to
be as secure as possible. Even in the case that someone might gain access to the underlying Postgres database, it should
not be able to break into the system.   
All important values are stored encrypted inside the database, while the encryption happens in-memory inside Nioca
itself. This means it should be able (after a 3rd party security audit maybe) to use Nioca even in an environment where
you could not even trust the DBA.

## How it works

Nioca implements zero trust all the way and it even does not trust the DBA by default.  
When it is started up for the very first time, you need to initialize it. You can use the same binary that starts the
server for generating X509 / SSH Certificate Authorities, full certificate chains (CA -> Intermediate -> EndEntity) and
just the EndEntity certificates themselves with the CLI.

1. For the very first start, you need to generate at least the Root + Intermediate CA on another host. The most optimal
   way of doing it would be to use an offline host with a Live OS booted into memory and saving the generated CA on an
   encrypted USB stick (for instance something like the iStorage datAshur). This might seem a bit paranoid, but that is
   the whole mindset behind Nioca's design. For testing, you can just do it on any host you like, of course.
2. After the full chain has been generated, you have all the files needed for the very first initialization.
   Just open your browser, and you will be redirected to Nioca's Initialization page where you can add all the necessary
   information.
3. The Master Shard Keys 1 + 2 must be saved from this stage. There is not a single way you can ever recover or see
   them again after this step. These keys will be used for the unsealing process.
4. When Nioca is started up (and initialized) it will be sealed by default. You need to unseal it with the 2 Master
   Key's
   from the initialization stage. From these Keys, Nioca will build up the master encryption key, which will then be
   used
   to decrypt all the other encryption keys and secrets from inside the database.
5. After Nioca has been unsealed, it can be used normally. There is only one single local root / admin user.
6. After your first login with the local root user, you can configure an SSO provider (like
   [rauthy](https://github.com/sebadob/rauthy) for instance). Daily operations should be done by Nioca Admins logged in
   via SSO.

# Prerequisites

Most things are built in and even database migrations between versions are embedded in the code.  
However, the very first database and user must exist. Create them on any postgres instance with
the following:

```sql
CREATE USER nioca WITH PASSWORD '123SuperSafe';
CREATE DATABASE nioca WITH OWNER nioca;
```

## Setup

1. Download the `nioca` binary from the `out/` directory for Linux, or (at this early stage of the project) clone this
   repo and build it yourself with `cargo build --release`

2. Generate the X509 full certificate chain and the unsealing certificates. Adjust the `--alt-name-ip` and
   `--alt-name-dns` to your needs.

```
./nioca x509 \
  --o 'My Org' \
  --alt-name-dns localhost \
  --alt-name-ip 127.0.0.1 \
  --alt-name-ip 192.168.14.50 \
  --usages digital-signature \
  --usages-ext server-auth \
  --stage full \
  --clean
```

- Enter your chosen password for the Root CA 3 times (new password, confirm, then test it afterward)
- Depending on your machine, these steps might take a while, because of a very heavy weight Argon2Id hashing behind
  the scenes to make it as secure as possible. You need at least 256MB of memory available for this operation!
- Enter your chosen password for the Intermediate CA 3 times. This should be different from the Root CA password!
- On success, you will have a full certificate chain generated in `ca/x509/`

3. To be able to start up Nioca, you will need a config first. Use `.env.deploy` from this repo as a minimal template.
   All values should be documented enough to understand, what you need to configure there.
   The `UNSEAL_CERT_B64` and `UNSEAL_KEY_B64` can be found in your newly generated CA chain folder:

- `UNSEAL_CERT_B64`: `cat ca/x509/end_entity/1/cert-chain.pem.b64`
- `UNSEAL_KEY_B64`: `cat ca/x509/end_entity/1/key.pem.b64`

4. Start the container:

- If you have a Postgres running on your localhost:

```
docker run --rm -v ./.env.deploy:/.env --network="host" ghcr.io/sebadob/nioca
```

- If you have it inside the same docker network, for instance with docker compose:

```
docker run --rm -v ./.env.deploy:/.env -p 8080:8080 -p 8443:8443 ghcr.io/sebadob/nioca
```

**Note:** Currently, this step is a bit tricky. To have a "secure" TLS connection in the browser, we MUST use port 443.
To make development simpler, this is using the `root` user inside the docker image for now. Another solution can be
found in the `justfile` where we grant a process the capabilities to bind to a port below 1024. A solution for this will
come in the future which makes it possible to run the container with a non-root user.

5. Click the link from inside the log output to get to the unseal / init UI
6. Fill out all the values, which you either know, like the encryption password for the Intermediate CA or the
   `INIT_KEY`, or you get them from the generated CA files:

- Root Certificate in PEM format: `cat ca/x509/root/root.cert.pem`
- Intermediate Certificate in PEM format: `cat ca/x509/intermediate/intermediate.cert.pem`
- Intermediate Key in encrypted PEM-HEX format: `cat ca/x509/intermediate/intermediate.key.pem.hex`
- Intermediate Key encryption password: The password for the Intermediate CA from the CA generation from step 2
- Nioca Init Key: The `INIT_KEY` from the `.env.deploy` file or the Nioca logs output
- The new password should be at least 16 characters. This is the local root user's password and grant access to
  everything.
- Click `Validate`
- If you entered all information correctly, you will see the parsed certificates and verify them before clicking
  `Initialize`
- **IMPORTANT**: Save the 2 master keys from this step in the most secure way you can think of! Preferable in 2 totally
  different locations with different access systems and / or credentials.  
  You will only see them at this stage and (at least for now) there is no way to restore or ever see them again! When
  you have saved the Master Key's, click `PROCEED`

7. After the initialization, you will see the default screen when Nioca is sealed. It will always be in this stage
   after a restart to make the whole system as tamper resistant as possible.
8. Enter the Master Keys from the initialization. The order does not matter. The by default 10 second rate limiter
   should be increased in production if possible to make it even more brute force resistant. This rate limiter is global
   and not bound to an IP or anything else. After a key has been entered, the whole Nioca application will not accept
   any
   other input from any source in the sealed state.
9. Depending on your system, the unsealing process could take a while because of the very heavy weight Argon2Id hashing.
10. After your first login, you can (and should) create an SSH CA from the navigation.
11. More in-depth readme and tutorials will follow in the future.

## Running behind an Ingress proxy

You can run Nioca behind an ingress (Traefik in this example) as well. A reason could be because you just do not have
multiple different IP addresses or no LoadBalancer in front of smaller setups.

There is currently no way (apart from using `DEV_MODE`) to start a plain HTTP server.
You should never(!) have any unencrypted connection to Nioca, since it has one of the most critical roles in your
infrastructure!  
The only situation in which you could actually start a plain HTTP server is, when you are running it behind an ingress,
and you are using a service mesh that actually wraps every single connection inside mTLS anyway already.
For this special use case, there will be an option in the future to actually start a plain HTTP server.
For now, you will need to add the Root Certificate to Traefik if you are running Nioca behind it. You will get an
Internal Server Error from Traefik, if you do not do this.

This is a very short gist to provide a starting point.

- set `PORT_HTTPS=8443` and `PORT_HTTPS_PUB=443` in Nioca's config and restart
- create a secret inside the `traefik` namespace or where ever your instance is running:
```yaml
kubectl -n traefik create secret generic root.cert.pem --from-file=root.cert.pem
```

- add a startup arg to Traefik: `--serversTransport.rootCAs=root.cert.pem`
- add a Kubernetes Service for Nioca like the following:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: nioca
  namespace: nioca
spec:
  selector:
    app: nioca
  ports:
    - name: https
      port: 8443
      targetPort: 8443
```
- Traefik IngressRoute definition:
```yaml
kind: Middleware
metadata:
  name: https-only
  namespace: nioca 
spec:
  redirectScheme:
    scheme: https
    permanent: true
---
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: https
  namespace: nioca 
spec:
  entryPoints:
    - websecure
  routes:
    - match: Host(`ca.exmplae.com`)
      kind: Rule
      services:
        - name: nioca
          port: 8443
---
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: http-redirect
  namespace: nioca
spec:
  entryPoints:
    - web
  routes:
    - match: Host(`ca.exmplae.com`)
      priority: 1
      kind: Rule
      middlewares:
        - name: https-only
      services:
        - name: nioca
          port: 8443
```


## Development

### sqlx compile time checked queries

If you are in DEV mode and the compilation fails because of the `sqlx::query!` macro, which complains about a not
yet applied migration (Chicken - Egg Problem), you need to install the sqlx cli to apply the migration manually beforehand.
The CLI will use the values from the `.env` file in the root folder.

`cargo install sqlx-cli`

You can then create and drop the database from the cli:

`sqlx database create`
`sqlx database drop`

And run migrations:

`sqlx migrate run`

Or a specific migration:<br>
`sqlx migrate add <name>`

# Issuing certificates via CLI - Examples

You can issue new intermediate or end entity certificates with the already existing and created Root CA.

To do so, you will need the secret keys for the private key decryption.
Without these keys, you will not be able to issue new certificates.
To create a new intermediate certificate, you need the Root CA encryption key.
To create a new end entity certificate, you need the Intermediate CA encryption key.

Note: Issuing multiple intermediate CA's is currently not supported by the current `nioca` CLI tool.
If you need multiple Intermediates, create a new folder and just copy the 'root' content over.

However, you can issue as many end entity certificates as you like. They will be created in sub folders.
The folder names will match their serial numbers.

### Example for issuing a new intermediate certificate:
```
nioca x509 \
    --cn 'My Common Name' \
    --c 'DE' \
    --l 'Dusseldorf' \
    --ou 'My Company - Nioca' \
    --o 'My Company' \
    --st 'NRW' \
    --stage intermediate
```

### Example for issuing an end entity certificate:
```
nioca x509 \
    --cn 'ca.example.com' \
    --c 'DE' \
    --l 'Dusseldorf' \
    --o 'My Company' \
    --st 'NRW' \
    --alt-name-dns 'ca.example.com' \
    --usages-ext server-auth \
    --usages-ext client-auth \
    --stage end-entity
```


CHANGELOG.md
# Changelog

## v0.4.4

This release just solves a small but annoying bug in the UI, where the deserialization for `EdDSA` X509
token algorithm could not be deserialized correctly.

## v0.4.3

- license change to Apache-2.0
- fix for the Dockerfile, where the multi-platform image screwed up the old behavior with Entrypoint + Cmd
- fix for wrong naming of EdDSA algorithm
- updated dependencies

## v0.4.2

- fully updated frontend and backend dependencies
- container image now available for `linux/amd64` and `linux/arm64`

## v0.4.1

This release only updates all dependencies.

## v0.4.0

This is the first open source release.


src/server.rs
use crate::certificates::encryption::{kdf_danger_static, EncAlg};
use crate::certificates::x509::end_entity::nioca_server_cert;
use crate::config::{Config, ConfigSealed, Db, EncKeys};
use crate::constants::{AUTO_UNSEAL, DEV_MODE, INSTANCE_UUID, XSRF_HEADER};
use crate::models::api::openapi::ApiDoc;
use crate::models::db::enc_key::EncKeyEntity;
use crate::routes::{ca, groups, unsealed, users};
use crate::routes::{clients_ssh, sealed};
use crate::routes::{clients_x509, oidc};
use crate::schedulers::scheduler_main;
use crate::service::password_hasher;
use crate::VERSION;
use axum::body::Bytes;
use axum::handler::HandlerWithoutStateExt;
use axum::headers::HeaderValue;
use axum::http::{header, HeaderName, StatusCode, Uri};
use axum::response::Redirect;
use axum::routing::{delete, get, post, put};
use axum::{extract, BoxError, Router};
use axum_server::tls_rustls::RustlsConfig;
use axum_server::Handle;
use base64::{engine::general_purpose, Engine as _};
use std::env;
use std::net::SocketAddr;
use std::str::FromStr;
use std::sync::Arc;
use std::time::Duration;
use tokio::time;
use tower::ServiceBuilder;
use tower_http::services::ServeDir;
use tower_http::trace::DefaultOnResponse;
use tower_http::{trace::TraceLayer, LatencyUnit, ServiceBuilderExt};
use tracing::{debug, info, warn};
use utoipa_swagger_ui::SwaggerUi;
use uuid::Uuid;
use x509_parser::nom::AsBytes;

#[derive(Debug, Clone, Copy)]
struct Ports {
    http: u16,
    https: u16,
    https_pub: u16,
}

impl Ports {
    fn build() -> Self {
        let http = env::var("PORT_HTTP")
            .unwrap_or_else(|_| "8080".to_string())
            .parse::<u16>()
            .expect("Error parsing PORT_HTTP to u16");
        let https = env::var("PORT_HTTPS")
            .unwrap_or_else(|_| "443".to_string())
            .parse::<u16>()
            .expect("Error parsing PORT_HTTPS to u16");
        let https_pub = env::var("PORT_HTTPS_PUB")
            .unwrap_or_else(|_| "443".to_string())
            .parse::<u16>()
            .expect("Error parsing PORT_HTTPS_PUB to u16");
        Self {
            http,
            https,
            https_pub,
        }
    }
}

pub async fn run_server(level: &str) -> Result<(), anyhow::Error> {
    info!("Nioca v{}", VERSION);
    info!("Log Level set to {}", level);

    info!("Nioca instance uuid: {}", *INSTANCE_UUID);

    Db::init().await?;

    // build middleware
    let sensitive_headers: Arc<[_]> = vec![
        header::AUTHORIZATION,
        header::COOKIE,
        HeaderName::from_str(XSRF_HEADER).unwrap(),
    ]
    .into();
    let middleware = ServiceBuilder::new()
        // Mark the `Authorization` and `Cookie` headers as sensitive so it doesn't show in logs
        .sensitive_request_headers(sensitive_headers.clone())
        // Add high level tracing / logging to all requests
        .layer(
            TraceLayer::new_for_http()
                .on_body_chunk(|chunk: &Bytes, latency: Duration, _: &tracing::Span| {
                    tracing::trace!(size_bytes = chunk.len(), latency = ?latency, "sending body chunk")
                })
                // .make_span_with(DefaultMakeSpan::new().include_headers(true))
                // .on_response(DefaultOnResponse::new().latency_unit(LatencyUnit::Micros)),
                .on_response(DefaultOnResponse::new().include_headers(true).latency_unit(LatencyUnit::Micros)),
        )
        .sensitive_response_headers(sensitive_headers)
        .compression()
        .append_response_header(
            header::X_FRAME_OPTIONS,
            HeaderValue::from_static("SAMEORIGIN")
        )
        .append_response_header(
            header::X_XSS_PROTECTION,
            HeaderValue::from_static("1;mode=block")
        )
        .append_response_header(
            header::X_CONTENT_TYPE_OPTIONS,
            HeaderValue::from_static("nosniff")
        )
        .append_response_header(
            header::STRICT_TRANSPORT_SECURITY,
            // HeaderValue::from_static("max-age=31536000; includeSubDomains")
            HeaderValue::from_static("max-age=31536000;includeSubDomains;preload")
        )
        .append_response_header(
            header::REFERRER_POLICY,
            HeaderValue::from_static("no-referrer")
        )
        .append_response_header(
            header::CONTENT_SECURITY_POLICY,
            // TODO add nonce's to inline scripts
            HeaderValue::from_static("default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; frame-ancestors 'none'; object-src 'none';")
        )
        .append_response_header(
            header::CACHE_CONTROL,
            HeaderValue::from_static("no-store")
        )
        .append_response_header(
            header::PRAGMA,
            HeaderValue::from_static("no-cache")
        );

    // redirect http -> https
    let ports = Ports::build();
    debug!("{ports:?}");
    if !*DEV_MODE {
        tokio::spawn(redirect_http_to_https(ports));
    }

    // run the password hasher
    tokio::spawn(password_hasher::run());

    // TLS config unseal
    let tls_config_unseal = {
        let cert = env::var("UNSEAL_CERT_B64").expect("UNSEAL_CERT_B64 is missing");
        let cert_bytes = general_purpose::STANDARD
            .decode(&cert)
            .expect("Error decoding UNSEAL_CERT_B64 from b64 to bytes");
        let key = env::var("UNSEAL_KEY_B64").expect("UNSEAL_KEY_B64 is missing");
        let key_bytes = general_purpose::STANDARD
            .decode(&key)
            .expect("Error decoding UNSEAL_KEY_B64 from b64 to bytes");
        RustlsConfig::from_pem(
            cert_bytes.as_bytes().to_vec(),
            key_bytes.as_bytes().to_vec(),
        )
        .await?
    };

    // after startup, everything is sealed -> start up special routes first
    let (tx_enc_key, rx_enc_keys) = flume::unbounded();
    let (tx_exit, rx_exit) = flume::unbounded();

    // to have a better quality of life in dev mode, we can auto-unseal
    if *AUTO_UNSEAL {
        tokio::spawn(auto_unseal(tx_enc_key.clone(), tx_exit.clone()));
    }

    let config_sealed = ConfigSealed::new(tx_enc_key, tx_exit).await?;
    let routes_sealed = Router::new()
        .route("/api/status", get(sealed::get_status))
        .nest(
            "/unseal",
            Router::new()
                .route("/execute", post(sealed::post_unseal))
                .route("/init", post(sealed::post_init))
                .route("/init/check", post(sealed::post_init_check))
                .route("/key", post(sealed::post_master_shard))
                .route("/status", get(sealed::get_status))
                .route("/xsrf", get(sealed::get_xsrf)),
        )
        .nest_service("/", ServeDir::new("static"))
        .layer(middleware.clone().into_inner())
        .with_state(config_sealed);
    debug!("Router created");

    let shutdown_handle = Handle::new();
    let handle_clone = shutdown_handle.clone();
    tokio::spawn(async move {
        let _ = rx_exit.recv_async().await;

        // shut down the sealed server
        time::sleep(Duration::from_secs(1)).await;
        handle_clone.shutdown();
        debug!("Shutdown signal for sealed server sent");
    });

    if *DEV_MODE {
        let addr = SocketAddr::from(([0, 0, 0, 0], ports.http));
        info!("Server listening on {}", addr);
        axum_server::bind(addr)
            .handle(shutdown_handle)
            .serve(routes_sealed.into_make_service())
            .await
            .expect("Starting the axum sealed server");
    } else {
        let addr = SocketAddr::from(([0, 0, 0, 0], ports.https));
        info!("Server listening on {}", addr);
        axum_server::bind_rustls(addr, tls_config_unseal)
            .handle(shutdown_handle)
            .serve(routes_sealed.into_make_service())
            .await
            .expect("Starting the axum sealed server");
    }

    let enc_keys = rx_enc_keys.recv_async().await?;

    // The sealed server has shut down, port is free again
    let app_state = Config::new(enc_keys)
        .await
        .expect("Building AppState Config - Is Nioca initialized?");

    // start OIDC SSO cache handler
    // TODO

    // start schedulers
    tokio::spawn(scheduler_main(app_state.clone()));

    // TLS config main -> With the intermediate certificates from the DB we now generate our own
    let tls_config = {
        let ca_cert = &app_state.read().await.nioca_signing_cert;
        let ca_chain = &app_state.read().await.ca_chain_pem;
        let (cert, key) = nioca_server_cert(ca_cert, ca_chain).await?;
        RustlsConfig::from_pem(cert.as_bytes().to_vec(), key.as_bytes().to_vec()).await?
    };

    // main routes
    let routes = Router::new()
        .nest(
            "/api",
            Router::new()
                .route("/ca/ssh", get(ca::get_ca_ssh))
                .route("/ca/ssh/external", post(ca::post_external_ca_ssh))
                .route("/ca/ssh/generate", post(ca::post_generate_ca_ssh))
                .route("/ca/ssh/:id", delete(ca::delete_ca_ssh))
                .route("/ca/x509", get(ca::get_ca_x509).post(ca::post_ca_x509))
                .route("/ca/x509/inspect", get(ca::get_ca_x509_inspect))
                .route("/ca/x509/:id", delete(ca::delete_ca_x509))
                .route(
                    "/clients/ssh",
                    get(clients_ssh::get_clients).post(clients_ssh::post_client),
                )
                .route(
                    "/clients/ssh/:id",
                    get(clients_ssh::get_client)
                        .put(clients_ssh::put_client)
                        .delete(clients_ssh::delete_client),
                )
                .route(
                    "/clients/ssh/:id/cert",
                    post(clients_ssh::post_build_client_cert),
                )
                .route(
                    "/clients/ssh/:id/secret",
                    get(clients_ssh::get_client_secret).put(clients_ssh::put_client_secret),
                )
                .route(
                    "/clients/x509",
                    get(clients_x509::get_clients).post(clients_x509::post_client),
                )
                .route(
                    "/clients/x509/:id",
                    get(clients_x509::get_client)
                        .put(clients_x509::put_client)
                        .delete(clients_x509::delete_client),
                )
                .route(
                    "/clients/x509/:id/cert",
                    post(clients_x509::post_build_client_cert),
                )
                .route(
                    "/clients/x509/:id/cert/p12",
                    post(clients_x509::post_build_client_cert_p12),
                )
                .route(
                    "/clients/x509/:id/secret",
                    get(clients_x509::get_client_secret).put(clients_x509::put_client_secret),
                )
                .route("/groups", get(groups::get_groups).post(groups::post_group))
                .route(
                    "/groups/:id",
                    put(groups::put_group).delete(groups::delete_group),
                )
                .route("/login", post(unsealed::post_login))
                .route("/login/check", get(unsealed::get_login_check))
                .route("/logout", post(unsealed::post_logout))
                .route("/password_change", put(unsealed::put_password_change))
                .route("/sessions", post(unsealed::post_session))
                .route("/oidc/auth", get(oidc::get_oidc_auth))
                .route("/oidc/auth/redirect", get(oidc::get_oidc_auth_redirect))
                .route("/oidc/callback", get(oidc::get_oidc_callback))
                .route(
                    "/oidc/config",
                    get(oidc::get_config_oidc).put(oidc::put_config_oidc),
                )
                .route("/oidc/exists", get(oidc::get_oidc_exists))
                .route("/status", get(unsealed::get_status))
                .route("/users", get(users::get_users))
                .route("/users/:id/access", get(users::get_user_group_access))
                .route(
                    "/users/:user_id/access/:group_id",
                    post(users::post_user_group_access)
                        .put(users::put_user_group_access)
                        .delete(users::delete_user_group_access),
                ),
        )
        .route("/unseal/status", get(unsealed::get_status))
        .route("/root.fingerprint", get(unsealed::get_root_fingerprint))
        .route("/root.pem", get(unsealed::get_root_pem))
        .merge(SwaggerUi::new("/docs/swagger-ui").url("/docs/openapi.json", ApiDoc::build()))
        .nest_service("/", ServeDir::new("static"))
        .layer(middleware.into_inner())
        .with_state(app_state);

    if *DEV_MODE {
        let addr = SocketAddr::from(([0, 0, 0, 0], ports.http));
        info!("Server listening on {}", addr);
        axum_server::bind(addr)
            .serve(routes.into_make_service())
            .await
            .expect("Starting the axum server");
    } else {
        let addr = SocketAddr::from(([0, 0, 0, 0], ports.https));
        info!("Server listening on {}", addr);
        axum_server::bind_rustls(addr, tls_config)
            .serve(routes.into_make_service())
            .await
            .expect("Starting the axum server");
    }

    Ok(())
}

async fn redirect_http_to_https(ports: Ports) {
    fn make_https(host: String, uri: Uri, ports: Ports) -> Result<Uri, BoxError> {
        let mut parts = uri.into_parts();

        parts.scheme = Some(axum::http::uri::Scheme::HTTPS);

        if parts.path_and_query.is_none() {
            parts.path_and_query = Some("/".parse().unwrap());
        }

        let https_host = host.replace(&ports.http.to_string(), &ports.https_pub.to_string());
        parts.authority = Some(https_host.parse()?);

        Ok(Uri::from_parts(parts)?)
    }

    let redirect = move |extract::Host(host): extract::Host, uri: Uri| async move {
        match make_https(host, uri, ports) {
            Ok(uri) => Ok(Redirect::permanent(&uri.to_string())),
            Err(error) => {
                tracing::warn!(%error, "failed to convert URI to HTTPS");
                Err(StatusCode::BAD_REQUEST)
            }
        }
    };

    let addr = SocketAddr::from(([0, 0, 0, 0], ports.http));
    debug!("Https redirect listening on {}", addr);

    axum::Server::bind(&addr)
        .serve(redirect.into_make_service())
        .await
        .expect("Binding HTTP redirect service to configure HTTP_PORT");
}

async fn auto_unseal(tx: flume::Sender<EncKeys>, tx_exit: flume::Sender<()>) {
    warn!(
        r#"

        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        !!! Auto-Unsealing is activated - DO NOT USE IN PRODUCTION !!!
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        "#
    );

    let master_shard_1 = env::var("AUTO_UNSEAL_SHARD_1").expect("AUTO_UNSEAL_SHARD_1");
    let master_shard_2 = env::var("AUTO_UNSEAL_SHARD_2").expect("AUTO_UNSEAL_SHARD_2");
    let master_full = format!("{}{}", master_shard_1, master_shard_2);
    // let master_key_digest = digest::digest(&digest::SHA256, master_full.as_bytes());
    let master_key = kdf_danger_static(master_full.as_bytes())
        .await
        .expect("Bad DEV AUTO_UNSEAL setup");
    // this is our master key for decryption end enc keys
    // let master_key = master_key_hash.as_ref().to_vec();

    let enc_id = env::var("AUTO_UNSEAL_ENC_UUID").expect("AUTO_UNSEAL_ENC_UUID");
    let enc_hex = env::var("AUTO_UNSEAL_ENC_VALUE").expect("AUTO_UNSEAL_ENC_VALUE");
    let enc_bytes = hex::decode(enc_hex).expect("AUTO_UNSEAL_ENC_VALUE decoding");

    let enc_keys = EncKeys {
        master_shard_1: None,
        master_shard_2: None,
        master_key,
        pepper: master_full,
        enc_key: EncKeyEntity {
            id: Uuid::from_str(&enc_id).expect("AUTO_UNSEAL_ENC_UUID to uuid"),
            alg: EncAlg::ChaCha20Poly1305.to_string(),
            value: enc_bytes,
        },
    };

    tx.send_async(enc_keys).await.unwrap();
    time::sleep(Duration::from_millis(100)).await;
    tx_exit.send_async(()).await.unwrap();
}
